### DISCUSSION TITLE
feat(ListView): 添加分组排序回调

### DISCUSSION THREAD

**Discussion by densen2014** [dotnetcore/BootstrapBlazor#3683] [Q&A] [answered]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Is your feature request related to a problem? Please describe the problem.

使用分组后没有采用排序属性

### Describe the solution you'd like

需要添加一个自定义排序

**Comment by bb-auto** [comment]
@densen2014 Thank you for contacting us. We will give feedback later.

**Comment by densen2014** [comment]
例子

[ConsoleApp1.zip](https://github.com/user-attachments/files/15894070/ConsoleApp1.zip)

版本2

[ConsoleApp1版本2.zip](https://github.com/user-attachments/files/15894420/ConsoleApp1.2.zip)


执行结果

![image](https://github.com/dotnetcore/BootstrapBlazor/assets/8428709/06fd2794-abdc-4461-954f-1f020a15db18)


生产数据

<details>
我程序用到的分组

![image](https://github.com/dotnetcore/BootstrapBlazor/assets/8428709/62cae656-c02a-4b0f-a223-9a19869c2244)

![image](https://github.com/dotnetcore/BootstrapBlazor/assets/8428709/2e41531f-3ec9-49a5-b3bd-32410fe8236e)

期望的效果

var groups2 = OrdersMenu.GroupBy(a => new {a.PrintGroup, a.CategoryName }).OrderBy(a => a.Key.PrintGroup).ThenBy(a => a.Key.CategoryName);

var groups2Name= groups2.Select(a => a.Key.CategoryName).ToList();



![image](https://github.com/dotnetcore/BootstrapBlazor/assets/8428709/22b0fcc6-e7e9-4bde-a52f-aa15b1426f43)

![image](https://github.com/dotnetcore/BootstrapBlazor/assets/8428709/b02854bd-e15f-4020-977f-c74145ba4b22)

![image](https://github.com/dotnetcore/BootstrapBlazor/assets/8428709/a74e10ed-9552-40db-83d5-b168b8666254)

</details>

**Reply by ArgoZhang** [reply, marked as answer]
![image](https://github.com/dotnetcore/BootstrapBlazor/assets/22001478/fd87c83d-67af-4f42-9030-0ecf6cab6af2)

### 定义分组类
```
record ProductGroupKey
{
    public int PrintGroup { get; set; }

    public string Category { get; set; } = "";

    public override string ToString()
    {
        return $"{PrintGroup}-{Category}";
    }
}
```

### 定义分组回调参数
`GroupName="p => new ProductGroupKey() { PrintGroup = p.PrintGroup, Category = p.Category }"`

```
<ListView TItem="Product" GroupName="p => new ProductGroupKey() { PrintGroup = p.PrintGroup, Category = p.Category }" GroupOrderCallback="GroupOrderCallback" GroupItemOrderCallback="GroupItemOrderCallback" OnQueryAsync="@OnQueryAsync">
    <BodyTemplate>
        <Card>
            <BodyTemplate>
                <div class="lv-demo-desc">@context.Category</div>
            </BodyTemplate>
        </Card>
    </BodyTemplate>
</ListView>
```

### 定义分组排序回调
```
private static IOrderedEnumerable<IGrouping<object?, Product>>? GroupOrderCallback(IEnumerable<IGrouping<object?, Product>> group)
{
    return group.OrderBy(i => ((ProductGroupKey)i.Key!).PrintGroup).ThenBy(i => ((ProductGroupKey)i.Key!).Category);
}
```

完整工程见附件
[BootstrapBlazorApp1.zip](https://github.com/user-attachments/files/15894589/BootstrapBlazorApp1.zip)

**Comment by densen2014** [comment]
给老张头点赞

**Comment by densen2014** [comment]
tips: 最终采用重写排序方法搞定, 因为区域排序不同导致中英文排序跟预想不符

```
    public int CompareTo(CategoryGroup other)
    {
        return string.Compare((PrintGroup??0).ToString() + CategoryName, (other.PrintGroup??0).ToString() + other.CategoryName, StringComparison.Ordinal);
    }

    public static IOrderedEnumerable<IGrouping<object?, ResOrdersMenu>>? GroupOrderCallback(IEnumerable<IGrouping<object?, ResOrdersMenu>> group)
    {
        return group.OrderBy(i => (CategoryGroup)i.Key!); 
    }
```
